### PR TITLE
[USBOHCI_NEW] Bring-in the USB OHCI miniport driver created by Vadim Galyant.

### PR DESCRIPTION
**Pull request by AmineKhaldi**
I kept the module enabled temporarily for the CIs. I'll introduce a commit that disables it once the builds are confirmed.

### CODE REVIEW

**Comment by binarymaster** [comment]
Commit author needs to be fixed, there is nickname now.

**Comment by AmineKhaldi** [comment]
@binarymaster fixed for both PRs.

File: drivers/usb/usbohci_new/usbohci.c [hunk 1]
```diff
@@ -0,0 +1,2724 @@
...
+
+    OhciHCCA = &OhciExtension->HcResourcesVA->HcHCCA;
+    DPRINT_OHCI("OHCI_InitializeSchedule: OhciHCCA - %p\n", OhciHCCA);
+
+    /* Set head pointers which start from HCCA */
+    for (ix = 0, jx = (INTERRUPT_ENDPOINTs - ENDPOINT_INTERRUPT_32ms);
+         ix < OHCI_NUMBER_OF_INTERRUPTS;
+         ix++, jx++)
+    {
+        static const UCHAR Balance[OHCI_NUMBER_OF_INTERRUPTS] =
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.c]: global please.

File: drivers/usb/usbohci_new/usbohci.c [hunk 2]
```diff
@@ -0,0 +1,2724 @@
...
+    POHCI_ENDPOINT OhciEndpoint = ohciEndpoint;
+    POHCI_HCD_ED ED;
+
+    DPRINT_OHCI("OHCI_ReopenEndpoint: ... \n");
+
+    ED = OhciEndpoint->HcdED;
+
+    RtlCopyMemory(&OhciEndpoint->EndpointProperties,
+                  EndpointProperties,
+                  sizeof(USBPORT_ENDPOINT_PROPERTIES));
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.c]: `sizeof(OhciEndpoint->EndpointProperties)`

File: drivers/usb/usbohci_new/usbohci.c [hunk 3]
```diff
@@ -0,0 +1,2724 @@
...
+    ULONG BufferEnd;
+    UCHAR DataToggle;
+
+    DPRINT_OHCI("OHCI_ControlTransfer: Ext %p, Endpoint %p\n",
+                OhciExtension,
+                OhciEndpoint);
+
+    MaxTDs = OHCI_RemainTDs(OhciExtension, OhciEndpoint);
+
+    if (SGList->SgElementCount + 2 > MaxTDs)
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.c]: Magic value?

File: drivers/usb/usbohci_new/usbohci.c [hunk 4]
```diff
@@ -0,0 +1,2724 @@
...
+    FirstTD->OhciTransfer = (ULONG)OhciTransfer;
+
+    FirstTD->HwTD.Padded[0] = 0;
+    FirstTD->HwTD.Padded[1] = 0;
+
+    OhciTransfer->PendingTDs++;
+
+    RtlCopyMemory(&FirstTD->HwTD.SetupPacket,
+                  &TransferParameters->SetupPacket,
+                  sizeof(USB_DEFAULT_PIPE_SETUP_PACKET));
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.c]: `sizeof(FirstTD->HwTD.SetupPacket)`

File: drivers/usb/usbohci_new/usbohci.c [hunk 5]
```diff
@@ -0,0 +1,2724 @@
...
+
+    FirstTdPA = (POHCI_HCD_TD)FirstTD->PhysicalAddress;
+
+    FirstTD->HwTD.gTD.CurrentBuffer = (ULONG)&FirstTdPA->HwTD.SetupPacket;
+
+    BufferEnd = (ULONG)&FirstTdPA->HwTD.SetupPacket +
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.c]: `ULONG_PTR`

File: drivers/usb/usbohci_new/usbohci.h [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+  };
+} OHCI_HW_TRANSFER_DESCRIPTOR, *POHCI_HW_TRANSFER_DESCRIPTOR;
+
+C_ASSERT(sizeof(OHCI_HW_TRANSFER_DESCRIPTOR) == 32);
+
+typedef struct _OHCI_HCD_TD {
+  // Hardware part
+  OHCI_HW_TRANSFER_DESCRIPTOR HwTD; // must be aligned to a 32-byte boundary
+  // Software part
+  ULONG PhysicalAddress;
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.h]: This is always 32 bits, even on x64?

File: drivers/usb/usbohci_new/usbohci.h [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+  ULONG Flags;
+  ULONG OhciTransfer;
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.h]: This too?

File: drivers/usb/usbohci_new/hardware.h [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+  };
+  ULONG  AsULONG;
+} OHCI_TRANSFER_CONTROL, *POHCI_TRANSFER_CONTROL;
+
+C_ASSERT(sizeof(OHCI_TRANSFER_CONTROL) == sizeof(ULONG));
+
+typedef struct _OHCI_TRANSFER_DESCRIPTOR { // must be aligned to a 16-byte boundary
+  OHCI_TRANSFER_CONTROL Control;
+  ULONG CurrentBuffer; // physical address of the next memory location
+  ULONG NextTD; // pointer to the next TD on the list of TDs
```
Comment (learn-more) [inline drivers/usb/usbohci_new/hardware.h]: This and `CurrentBuffer` are always 32bit pointer?
Even on x64?

File: drivers/usb/usbohci_new/usbohci.c [hunk 9]
```diff
@@ -0,0 +1,2724 @@
...
+            TD->HwTD.gTD.Control.BufferRounding = 0;
+            TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_IN;
+        }
+        else
+        {
+            TD->HwTD.gTD.Control.BufferRounding = TRUE;
+            TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_OUT;
+        }
+
+        TD->HwTD.gTD.CurrentBuffer = 0;
```
Comment (learn-more) [inline drivers/usb/usbohci_new/usbohci.c]: If these are pointers, i'd prefer `NULL`

**Comment by yagoulas** [comment]
@AmineKhaldi : You can keep it in build but remove it from the cd

File: drivers/usb/usbohci_new/hardware.h (hunk 8 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/hardware.h]: Yes, only 32-bit pointers are always used for OHCI and UHCI.

File: drivers/usb/usbohci_new/usbohci.c [hunk 10]
```diff
@@ -0,0 +1,2724 @@
...
+NTAPI
+OHCI_OpenInterruptEndpoint(IN POHCI_EXTENSION OhciExtension,
+                           IN PUSBPORT_ENDPOINT_PROPERTIES EndpointProperties,
+                           IN POHCI_ENDPOINT OhciEndpoint)
+{
+    UCHAR Period;
+    ULONG PeriodIdx = 0;
+    POHCI_HCD_ED ED;
+    ULONG ScheduleOffset;
+    const UCHAR Index[8] = {
```
[the remark this answers is not in the document]
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Changed.

File: drivers/usb/usbohci_new/usbohci.c [hunk 11]
```diff
@@ -0,0 +1,2724 @@
...
+
+    ASSERT(Period != 0);
+
+    while (!(Period & 1))
+    {
+        PeriodIdx++;
+        Period >>= 1;
+    }
+
+    ASSERT(PeriodIdx < (sizeof(Index) / sizeof(Index[0])));
```
[the remark this answers is not in the document]
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Changed.

File: drivers/usb/usbohci_new/usbohci.c [hunk 12]
```diff
@@ -0,0 +1,2724 @@
...
+    POHCI_EXTENSION OhciExtension = ohciExtension;
+    POHCI_ENDPOINT OhciEndpoint = ohciEndpoint;
+    ULONG TransferType;
+    MPSTATUS MPStatus;
+
+    DPRINT_OHCI("OHCI_OpenEndpoint: ... \n");
+
+    RtlCopyMemory(&OhciEndpoint->EndpointProperties,
+                  EndpointProperties,
+                  sizeof(USBPORT_ENDPOINT_PROPERTIES));
```
[the remark this answers is not in the document]
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Changed.

File: drivers/usb/usbohci_new/usbohci.c (hunk 2 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Changed.

File: drivers/usb/usbohci_new/usbohci.c (hunk 1 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Changed.

File: drivers/usb/usbohci_new/usbohci.c (hunk 4 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Changed.

File: drivers/usb/usbohci_new/usbohci.h (hunk 6 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.h]: always 32

File: drivers/usb/usbohci_new/usbohci.h (hunk 7 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.h]: Yes

File: drivers/usb/usbohci_new/usbohci.c (hunk 5 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Changed.

File: drivers/usb/usbohci_new/usbohci.c (hunk 9 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: typedef struct _OHCI_TRANSFER_DESCRIPTOR { // must be aligned to a 16-byte boundary
...
  ULONG CurrentBuffer; // physical address of the next memory location
...
} OHCI_TRANSFER_DESCRIPTOR, *POHCI_TRANSFER_DESCRIPTOR;

File: drivers/usb/usbohci_new/usbohci.c (hunk 3 above)
Comment (vgalnt) [reply drivers/usb/usbohci_new/usbohci.c]: Add constant.

**Pull request merged**